nomt/src/beatree/mod.rs

```diff
@@ -626,19 +637,82 @@ impl AsyncLeafLoad {
+
+enum AsyncLookupState {
+    Initial(AsyncLeafLoad),
+    Overflow(AsyncOverflowReader, Option<usize>),
+    Done,
+}
 
 #[allow(dead_code)]
 impl AsyncLookup {
-    /// Finish the lookup.
-    /// Calling this with the wrong page will likely lead to panics or bugs in the future.
-    pub fn finish(self, page: FatPage) -> Option<Vec<u8>> {
-        let leaf = self.1.finish_inner(page);
-        ops::finish_lookup_blocking(self.0, &leaf, &self.1.read_tx.leaf_store)
+    /// Attempt to submit a continuation request along the handle.
+    ///
+    /// This should not be called unless `try_finish` has failed at least once.
+    pub fn submit(&mut self, io_handle: &IoHandle, user_data: u64) -> Option<OverflowMetadata> {
+        match self.state {
+            AsyncLookupState::Initial(_) => None,
+            AsyncLookupState::Overflow(ref mut overflow, _) => {
+                overflow.submit(io_handle, user_data).map(OverflowMetadata)
+            }
+            AsyncLookupState::Done => None,
+        }
+    }
+
+    /// Try to finish the lookup.
+    /// Calling this with the wrong page will lead to panics or silent errors.
+    ///
+    /// If calling for the first time, provide no metadata. Only provide metadata that has been
+    /// returned from `submit` from this lookup. Otherwise, this may panic or silently cause errors.
+    ///
+    /// This returns `None` if not finished, `Some` otherwise. After returning `Some` once, this
+    /// will return `None` forever.
+    ///
+    /// If more lookups are required to finish, this will return an `Err`.
+    pub fn try_finish(
+        &mut self,
+        page: FatPage,
+        meta: Option<OverflowMetadata>,
+    ) -> Option<Option<Vec<u8>>> {
+        match self.state {
+            AsyncLookupState::Done => return None,
+            AsyncLookupState::Initial(ref inner) => {
+                let leaf = inner.finish_inner(page);
+                match ops::finish_lookup_async(self.key, &leaf, &inner.read_tx.leaf_store) {
+                    Ok(val) => {
+                        self.state = AsyncLookupState::Done;
+                        return Some(val);
+                    }
+                    Err(overflow) => {
+                        self.state = AsyncLookupState::Overflow(overflow, None);
+                        return None;
+                    }
+                }
+            }
+            AsyncLookupState::Overflow(ref mut overflow, ref mut initial_meta) => {
+                // UNWRAP: part of function contract.
+                let index = meta
+                    .map(|m| m.0)
+                    .unwrap_or_else(|| initial_meta.take().unwrap());
+                let res = overflow.complete(index, page).map(Some);
+
+                if res.is_some() {
+                    self.state = AsyncLookupState::Done;
+                }
+
+                res
+            }
+        }
     }
 }
 
+/// Metadata related to an overflow page fetch.
+pub struct OverflowMetadata(usize);
```

Comment:
Not really clear what kind of metadata is that. Not even sure if that's a metadata, if you think about it.

Reply:
well, what do you recommend calling it?

Reply:
I don't really have a good name, not now. Something that mentions that it represents an inflight fetch and that it is an index etc